### ISSUE TITLE
Temporal support (TC39 stage 3 proposal)

### ISSUE THREAD

**Issue by notramo** [oven-sh/bun#15853]
### What is the problem this feature would solve?

The Temporal API is already supported by Deno and the Boa JS engine.

https://github.com/tc39/proposal-temporal

### What is the feature you are proposing to solve the problem?

Implement Temporal in Bun.

### What alternatives have you considered?

Slow and heavy polyfills.

**Comment by maurovanetti** [comment]
See https://github.com/temporalio/sdk-typescript/issues/1334 (_[Feature Request] Run temporal worker in bun_).

**Comment by kravetsone** [comment]
> See https://github.com/temporalio/sdk-typescript/issues/1334 (_[Feature Request] Run temporal worker in bun_).

Not the same temporal

**Comment by maurovanetti** [comment]
> > See [temporalio/sdk-typescript#1334](https://github.com/temporalio/sdk-typescript/issues/1334) (_[Feature Request] Run temporal worker in bun_).
> 
> Not the same temporal

This is embarrassing. Sorry. 😊

**Comment by TomasHubelbauer** [comment]
This would be great. It looks as though in Safari whose JavaScript engine, JavaScriptCore, Bun is using, has preview support with the `useTemporal` runtime flag: https://bugs.webkit.org/show_bug.cgi?id=223166 & https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Temporal#browser_compatibility

**Comment by notramo** [comment]
The patches contain just a stub, only 2 empty objects.

**Comment by kravetsone** [comment]
It would be great to have it
-1 dependency

**Comment by NuckChorris** [comment]
Looking upstream in JSC, despite the issue showing no movement in years, I see they've got an experimental implementation of Temporal: https://github.com/search?q=repo%3AWebKit%2FWebKit%20Temporal&type=code

I wonder if Bun could enable this experimentally like Deno did with the V8 Temporal implementation 👀

**Comment by Jarred-Sumner** [comment]
@NuckChorris try `BUN_JSC_useTemporal=1`: https://github.com/oven-sh/WebKit/blob/50ae7bfd503f896681ca5a9b848f9822c31c4912/Source/JavaScriptCore/runtime/OptionsList.h#L620

**Comment by notramo** [comment]
@Jarred-Sumner, isn't it only for the empty stub `Temporal` object? Does it enable a usable implementation?

**Comment by NuckChorris** [comment]
@notramo I would recommend running the [Temporal Test262 Suite](https://github.com/js-temporal/temporal-test262-runner/) against it (if I find time this weekend I might run it and report back) but from the C++ it looks at least vaguely implemented-ish?

**Comment by jacksteamdev** [comment]
<img width="419" alt="Image" src="https://github.com/user-attachments/assets/ad51301e-d557-418d-998c-291cf0b9ac29" />

It's about 40% there... but the tests run blazing fast! The test runner README estimates ~30s to run, but Bun runs it in 0.8s on my M3.

**Comment by kravetsone** [comment]
> | Implementation | Pass Rate | Execution Time | Notes |
> |----------------|-----------|----------------|-------|
> | `BUN_JSC_useTemporal` | 40% | 0.8s | |
> | `temporal-polyfill` | 88% | 6s | |
> | `@js-temporal/polyfill` | 84% | 228s | Some >10s Duration tests
> 
> It's about 40% there... but the tests run blazing fast! The test runner README estimates ~30s to run, but Bun runs it in 0.8s on my M3.
> 
> <img width="419" alt="Image" src="https://github.com/user-attachments/assets/ad51301e-d557-418d-998c-291cf0b9ac29" />
> 
> 

Wants it as much as possible sooner

Because it's really great api

**Comment by kravetsone** [comment]
https://github.com/nodejs/node/pull/57128

NodeJS wants to unflag it

**Comment by jasnell** [comment]
Yes, we'd like to unflag it in node.js but still keep it marked as experimental, at least until v8's test coverage numbers are close to 100 and v8 chooses to expose it by default without the flag.

**Comment by jasnell** [comment]
The current thinking is that we likely will not actually merge the PR until after 24.0.0 is cut in April and keep it as experimental until v8's test coverage finally catches up.

**Comment by odicho** [comment]
Thanks! I hope the team can afford to prioritize this, Temporal is an incredible advantage to have on your runtime. I have to stick with node until then, but I'm excited for this to land in bun!
